runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -2178,69 +2231,134 @@ bool TR_EscapeAnalysis::checkOtherDefsOfLoopAllocation(TR::Node *useNode, Candid
       if (trace())
          traceMsg(comp(), "      Look at def node [%p] for use node [%p]\n", defNode, useNode);
 
+      // _aliasesOfAllocNode contains sym refs that are just aliases for a fresh allocation
+      // i.e. it is just a simple attempt at tracking allocations in cases such as :
+      // ...
+      // a = new A()
+      // ...
+      // b = a
+      // ...
+      // c = b
+      //
+      // In this case a, b and c will all be considered aliases of an alloc node and so a load of
+      // any of those sym refs will be treated akin to how the fresh allocation would have been in the below logic
+      //
+
       bool allnewsonrhs = false;
-      if ((_valueNumberInfo->getValueNumber(defNode) == _valueNumberInfo->getValueNumber(candidate->_node)))
+
+      if ((defNode->getFirstChild() == candidate->_node) &&
+          (_valueNumberInfo->getValueNumber(defNode) == _valueNumberInfo->getValueNumber(useNode)))
+         allnewsonrhs = true;
+      else if ((_valueNumberInfo->getValueNumber(defNode) == _valueNumberInfo->getValueNumber(candidate->_node)) &&
+               (_useDefInfo->getTreeTop(defIndex)->getEnclosingBlock() == candidate->_block) &&
+               _aliasesOfAllocNode->get(defNode->getSymbolReference()->getReferenceNumber()))
+         allnewsonrhs = true;
+      else
          {
-         if ((defNode->getFirstChild() == candidate->_node) &&
-             (_valueNumberInfo->getValueNumber(defNode) == _valueNumberInfo->getValueNumber(useNode)))
-            allnewsonrhs = true;
-         else
+         allnewsonrhs = true;
+         TR_UseDefInfo::BitVector defs2(comp()->allocator());
+         _useDefInfo->getUseDef(defs2, useIndex);
+         TR_UseDefInfo::BitVector::Cursor cursor2(defs2);
+
+         // Loop over definitions for this use and over all the candidate
+         // allocations.  If the definition comes directly from a candidate
+         // for stack allocation, it's harmless; if it's copied from a
+         // variable that's aliased with a candidate for stack allocation
+         // that was allocated in the same block, it's harmless
+         for (cursor2.SetToFirstOne(); cursor2.Valid(); cursor2.SetToNextOne())
             {
-            allnewsonrhs = true;
-            TR_UseDefInfo::BitVector defs2(comp()->allocator());
-            _useDefInfo->getUseDef(defs2, useIndex);
-            TR_UseDefInfo::BitVector::Cursor cursor2(defs2);
-            for (cursor2.SetToFirstOne(); cursor2.Valid(); cursor2.SetToNextOne())
+            int32_t defIndex2 = cursor2;
+            if (defIndex2 == 0)
                {
-               int32_t defIndex2 = cursor2;
-               if (defIndex2 == 0)
+               allnewsonrhs = false;
+               break;
+               }
+
+            TR::Node *defNode2 = _useDefInfo->getNode(defIndex2);
+            TR::Node *firstChild = defNode2->getFirstChild();
+            bool rhsIsHarmless = false;
```

Comment:
could we have a comment on what harmless means in this context to provide the intuition for all the checks that follow?

Reply:
Addressed this in commit aa48bf3fd7a279f3feb25aba345b7ff6850f3c8f